src/Sarif.Viewer.VisualStudio/Telemetry/Events.cs

```diff
@@ -0,0 +1,76 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Sarif.Viewer.Telemetry
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Contains the methods for firing telemetry events from the SARIF viewer extension.
+    /// </summary>
+    /// <remarks>
+    /// The documentation in this class is super useful for understanding where this telemetry is fired.
+    /// The documentation should also contain "why" the telemetry is being fired, i.e., why is this data needed?
+    /// Also, by having individual methods, versus calling application insights directly from the code,
+    /// it allows us to easily locate the code that is firing the telemetry.
+    /// </remarks>
+    internal class Events
+    {
+        /// <summary>
+        /// Does nothing.
+        /// </summary>
+        /// <remarks>
+        /// The reason we call track event directly here is that
+        /// we are in the middle of MEF composition\creation <see cref="TelemetryProvider.TelemetryProviderService"/>, if we called
+        /// <see cref="TelemetryProvider.TrackEvent(string)"/> that would end up being a recursive MEF composition call (and a
+        /// MEF composition exception). This method is purely used for event naming purposes.
+        /// </remarks>
+        public static void ExtensionLoaded() => throw new NotImplementedException();
+
+        /// <summary>
+        /// Tracks an event indicating that a SARF log file was run was loaded into the extension.
+        /// </summary>
+        /// <param name="toolName">The name of the tool that provided the SARIF log.</param>
+        public static void LogFileRunCreatedByToolName(string toolName) =>
+            TelemetryProvider.TrackEvent<Events>(new Dictionary<string, string> { { nameof(toolName), toolName } } );
+
+        /// <summary>
+        /// Tracks an event indicating that a document was opened as a result of navigation either through
+        /// a call to <see cref="ResultTextMarker.NavigateTo(bool, bool)"/> or <see cref="CodeLocationObject.NavigateTo(bool, bool)"/>.
+        /// </summary>
+        public static void TaskItemDocumentOpened() =>
+            TelemetryProvider.TrackEvent<Events>();
+
+        /// <summary>
+        /// Indicates that a SARIF log file was opened (imported) through the tools menu.
+        /// </summary>
+        /// <param name="toolName">The name of the tool that produced the original static analysis file.</param>
+        public static void LogFileOpenedByMenuCommand(string toolName) =>
+            TelemetryProvider.TrackEvent<Events>(new Dictionary<string, string> { { nameof(toolName), toolName } });
+
+        /// <summary>
+        /// Called when a call to <see cref="Services.CloseSarifLogService.CloseAllSarifLogs"/> is invoked.
+        /// </summary>
+        public static void CloseAllSarifLogsApiInvoked() =>
```

Comment:
I am concerned about scaling this boilerplate pattern. The majority of the members here are already like this. The comment doesn't tell me something I don't know. It also creates an additional member that needs to be named that isn't meaningfully different than the caller.

Reply:
All I can say is this is the technique we use in most of our extensions and it works very well. The comment should include "where" it is invoked and "why" is is invoked.

Right now I would say that there are no "whys" for any of these.

We shouldn't be firing telemetry for data we will never consume.

So, we have the option to check in with NO telemetry points until we understand "why" we want this data.

For example, it isn't interesting to know that someone called an API, or that something was executed unless you plan to take action based on that data. Windows (for example) takes a hard stance on this since telemetry takes not only user bandwidth but storage space on Azure.



---
In reply to: [505753537](https://github.com/microsoft/sarif-visualstudio-extension/pull/226#discussion_r505753537) [](ancestors = 505753537)